lib/keila_web/controllers/api_controller.ex

```diff
@@ -0,0 +1,231 @@
+    end
+  end
+
+  def show_contact(conn, %{"id" => id}) do
+    case Contacts.get_project_contact(project_id(conn), id) do
+      contact = %Contacts.Contact{} -> render(conn, "contact.json", %{contact: contact})
+      nil -> send_404(conn)
+    end
+  end
+
+  def update_contact(conn, %{"id" => id}) do
+    if Contacts.get_project_contact(project_id(conn), id) do
+      case Contacts.update_contact(id, conn.assigns.data) do
+        {:ok, contact} -> render(conn, "contact.json", %{contact: contact})
+        {:error, changeset} -> send_changeset_error(conn, changeset)
+      end
+    else
+      send_404(conn)
+    end
+  end
+
+  def delete_contact(conn, %{"id" => id}) do
+    Contacts.delete_project_contacts(project_id(conn), filter: %{"id" => id})
+
+    conn
+    |> put_status(204)
+  end
+
+  #
+  # Campaign functions
+  #
+
+  plug ApiNormalizer,
+       [normalize: [{:data, :campaign}]]
+       when action in [:create_campaign, :update_campaign, :schedule_campaign]
+
+  @spec index_campaigns(Plug.Conn.t(), Map.t()) :: Plug.Conn.t()
+  def index_campaigns(conn, _params) do
+    campaigns =
+      Mailings.get_project_campaigns(project_id(conn))
+      |> then(fn campaigns ->
+        count = Enum.count(campaigns)
+        %Keila.Pagination{data: campaigns, page: 0, page_count: 1, count: count}
+      end)
+
+    render(conn, "campaigns.json", %{campaigns: campaigns})
+  end
+
+  def create_campaign(conn, _params) do
+    case Mailings.create_campaign(project_id(conn), conn.assigns.data) do
+      {:ok, campaign} -> render(conn, "campaign.json", %{campaign: campaign})
+      {:error, changeset} -> send_changeset_error(conn, changeset)
+    end
+  end
+
+  def show_campaign(conn, %{"id" => id}) do
+    case Mailings.get_project_campaign(project_id(conn), id) do
+      campaign = %Mailings.Campaign{} -> render(conn, "campaign.json", %{campaign: campaign})
+      nil -> send_404(conn)
+    end
+  end
+
+  def update_campaign(conn, %{"id" => id}) do
+    project_id = project_id(conn)
+
+    with campaign = %Mailings.Campaign{} <- Mailings.get_project_campaign(project_id, id) do
+      settings_id = campaign.settings.id
+      params = put_in(conn.assigns.data, ["settings", "id"], settings_id)
+
+      case Mailings.update_campaign(id, params) do
+        {:ok, campaign} -> render(conn, "campaign.json", %{campaign: campaign})
+        {:error, changeset} -> send_changeset_error(conn, changeset)
+      end
+    else
+      _ -> send_404(conn)
+    end
+  end
+
+  def delete_campaign(conn, %{"id" => id}) do
+    Mailings.delete_project_campaigns(project_id(conn), [id])
+
+    conn
+    |> put_status(204)
+  end
+
+  def deliver_campaign(conn, %{"id" => id}) do
+    # TODO immediate feedback on missing sender or insufficient credits
+    campaign = Mailings.get_project_campaign(project_id(conn), id)
+
+    if campaign do
+      Mailings.deliver_campaign_async(campaign.id)
+      put_status(conn, 204)
```

Comment:
Maybe in this case return status code 202 to indicate that the request was accepted but will be processed.